### PR TITLE
remove pid file on exit

### PR DESCRIPTION
**Pull request by geyslan**
### 1. Explain what the PR does

    tracee: remove pid file on exit
    
    This also does some cosmetic changes to the code.

### 2. Explain how to test it

```
$ sudo ./dist/tracee --filter event=ptrace_code_injection
TIME             UID    COMM             PID     TID     RET              EVENT                     ARGS
```
then
```
ctrl+c
```
and watch:
```
$ find /tmp/tracee
/tmp/tracee
/tmp/tracee/out
```

### 3. Other comments

Fix: #3074

### CODE REVIEW

File: pkg/ebpf/tracee.go [hunk 1]
```diff
@@ -1580,11 +1581,14 @@ func (t *Tracee) Close() {
 			logger.Errorw("failed to clean containers module when closing tracee", "err", err)
 		}
 	}
-	err := t.cgroups.Destroy()
-	if err != nil {
+	if err := t.cgroups.Destroy(); err != nil {
```
Comment (geyslan) [inline pkg/ebpf/tracee.go:1584]: Just a cosmetic to align with other checks.

**Review by josedonizetti — CHANGES_REQUESTED** [review]
There is a problem of having the PID file internal to `Tracee{}` library, instead of part of the `main.go`, which should create the pid file as the first thing, and clean it up at the end of it. 

For anything using the pid file as a metric to know if tracee is alive or not (tracee-action does it, and k8s files might do it in the future), cleaning inside the lib, means those services will think tracee is not live and either stop docker or restart the container, even though there is logic to be execute after `t.Run()` like drain the chan of events, for example.

So if possible, maybe we could fix this problem here? Ideally we should remove the pid creation from inside the lib, and have it under `cmd/tracee.go`as the first and last thing before the process exits.

**Comment by geyslan** [comment]
@josedonizetti this 768b20d is a proposal based on your concerns. I couldn't extract it to main as the pidfile uses capture output path logic which is only set in Runner.Run(). But with that I was able to extract the creation and destruction of the pidfile to the highest execution path. WDYT?

**Comment by yanivagman** [comment]
> There is a problem of having the PID file internal to `Tracee{}` library, instead of part of the `main.go`, which should create the pid file as the first thing, and clean it up at the end of it.
> 
> For anything using the pid file as a metric to know if tracee is alive or not (tracee-action does it, and k8s files might do it in the future), cleaning inside the lib, means those services will think tracee is not live and either stop docker or restart the container, even though there is logic to be execute after `t.Run()` like drain the chan of events, for example.
> 
> So if possible, maybe we could fix this problem here? Ideally we should remove the pid creation from inside the lib, and have it under `cmd/tracee.go`as the first and last thing before the process exits.

I think this problem is bigger than what this PR comes to solve (https://github.com/aquasecurity/tracee/issues/3074#issuecomment-1538724544) and we should probably tackle it later. In the future, I believe the liveness/readiness probes can be implemented using our API server, and not through this pid file.

**Comment by josedonizetti** [comment]
@yanivagman Liveness/readiness will use the server, that is why we introduced the `--healthz` flag, tracee-action uses the pid, for starting for now, so removing is not a problem, though others could be using the pid, for the ideas mentioned here, or others ideas, and having its removal before the full stop of the flow might lead to misbehavior. The pid should not be tied to the internals of the lib, but to the start/exit of the process.

**Comment by rafaeldtinoco** [comment]
Issue is being mitigated by: https://github.com/aquasecurity/tracee/pull/3076/commits/6aee729310853c3c472d56525a4214bc26107345 at https://github.com/aquasecurity/tracee/pull/3076/ so tests won't fail because of `tracee.pid` leftovers.

**Comment by geyslan** [comment]
@josedonizetti if you agree with this solution let's merge it.

**Review by josedonizetti — APPROVED** [review]

File: pkg/utils/files.go [hunk 2]
```diff
@@ -30,6 +30,15 @@ func OpenAt(dir *os.File, relativePath string, flags int, perm fs.FileMode) (*os
...
 
+// RemoveAt is a wrapper function to the `unlinkat` syscall using golang types.
+func RemoveAt(dir *os.File, relativePath string, flags int) error {
+	if err := unix.Unlinkat(int(dir.Fd()), relativePath, flags); err != nil {
+		return errfmt.WrapError(err)
+	}
+
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (rafaeldtinoco) [reply pkg/utils/files.go:41]: cool. merge when you are good. already merged my PR.

**Pull request merged**